### PR TITLE
loopify impute steps

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
Related to https://github.com/tidymodels/recipes/pull/1156.

This PR does a small refactor to `step_impute_linear()`, `step_impute_knn()`, and `step_impute_bag()` to bring them up to the same standard as https://github.com/tidymodels/recipes/pull/1156.

### CODE REVIEW

File: R/impute_bag.R [hunk 1]
```diff
@@ -271,29 +271,30 @@ prep.step_bagimpute <- prep.step_impute_bag
...
-        rlang::warn("All predictors are missing; cannot impute")
-      } else {
-        pred_vals <- predict(object$models[[imp_var]], pred_data)
-        # For an ipred bug reported on 2021-09-14:
-        pred_vals <- cast(pred_vals, object$models[[imp_var]]$y)
-        new_data[missing_rows, imp_var] <- pred_vals
-      }
+  for (col_name in col_names) {
+    missing_rows <- !complete.cases(new_data[[col_name]])
+    if (!any(missing_rows)) {
```
Comment (simonpcouch) [inline R/impute_bag.R:285]: Nice. :)

File: R/impute_knn.R [hunk 2]
```diff
@@ -247,41 +247,43 @@ nn_pred <- function(index, dat) {
 
 #' @export
 bake.step_impute_knn <- function(object, new_data, ...) {
-  col_names <- purrr::map(object$columns, function(x) unname(x$x)) %>%
-    purrr::flatten_chr() %>%
-    unique()
-  check_new_data(col_names, object, new_data)
+  col_names <- purrr::map_chr(object$columns, "y")
+  all_cols <- unique(unlist(object$columns, recursive = TRUE))
```
Comment (simonpcouch) [inline R/impute_knn.R:251]: Why both `col_names` and `all_cols`? I had initially assumed this was to accommodate non-standard roles but that seems unrelated.

Comment (EmilHvitfeldt) [reply R/impute_knn.R:251]: `col_names` is the variables we are imputing, `all_cols` are the variables we are imputing + the variables we use to impute with

Comment (EmilHvitfeldt) [reply R/impute_knn.R:251]: And we need to make sure that all needed variables are passed to `check_new_data()`, but we just need to loop over the variables we are imputing
